src/Microsoft.Health.Fhir.Core/Features/Conformance/SystemConformanceProvider.cs

```diff
@@ -104,16 +112,59 @@ public void ConfigureOptionalCapabilities(Action<ListedCapabilityStatement> buil
 
         public void Dispose()
         {
-            _sem?.Dispose();
-            _sem = null;
+            _defaultCapabilitySemaphore?.Dispose();
+            _defaultCapabilitySemaphore = null;
+            _metadataSemaphore?.Dispose();
+            _metadataSemaphore = null;
+        }
+
+        public async Task Handle(RebuildCapabilityStatement notification, CancellationToken cancellationToken)
+        {
+            await _metadataSemaphore.WaitAsync(cancellationToken);
+            try
+            {
+                if (_builder != null)
+                {
+                    switch (notification.Part)
+                    {
+                        case RebuildPart.SearchParameter:
+                            // Update search params;
+                            _builder.SyncSearchParameters();
+                            break;
+
+                        case RebuildPart.Profiles:
+                            // Update supported profiles;
+                            _builder.SyncProfiles(true);
+                            break;
+                    }
+                }
+
+                _metadata = null;
+            }
+            finally
+            {
+                _metadataSemaphore.Release();
+            }
         }
 
-        public Task Handle(SearchParametersUpdated notification, CancellationToken cancellationToken)
+        public override async Task<ResourceElement> GetMetadata(CancellationToken cancellationToken = default)
         {
-            // If search parameters change, rebuild the capability statement
-            _listedCapabilityStatement = null;
+            await _metadataSemaphore.WaitAsync(cancellationToken);
+            try
+            {
+                if (_metadata == null)
+                {
+                    _ = await GetCapabilityStatementOnStartup(cancellationToken);
 
-            return Task.CompletedTask;
+                    _metadata = _builder.Build().ToResourceElement();
```

Comment:
just for my understanding, the `_listedCapabilityStatement` now no longer contains the latest capability statement. All the updates are now made to capability statement "copy" inside the `CapabilityStatementBuilder`. Is there a reason we want to maintain two versions like that?

Reply:
I see capability statement like object which has dynamic and constant parts. Constant parts get populated once during start up and we do multiple checks against them via `IConformanceProvider.SatisfiesAsync`, dynamic parts can be changed at any time. 
Since constant part is constant we don't have to generate it all the time, so it's get initialized once via `_listedCapabilityStatement ` and just get queired all the time.
`_metadata` can be different so we reset it in `Handle` and build again in `GetMetadata` method.

We use same builder object as optimization, so we don't have to create constant part second time for metadata. 


---
In reply to: [611313190](https://github.com/microsoft/fhir-server/pull/1795#discussion_r611313190) [](ancestors = 611313190)

Reply:
Thanks for the explanation!